======================
The ``lazy_matrix_mul`` module
======================

Write a function that multiplies 2 matrices by using the module NumPy:

      Read: Matrix multiplication - only Matrix product (two matrices)
      	    Prototype: def lazy_matrix_mul(m_a, m_b):

	    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

Documentation
=============
Here we verify if our modules an funtions are documented, for holberton requirements is a must.
we use __doc__ to verify module's and function's Documentation

    >>> module_doc = __import__("101-lazy_matrix_mul").__doc__
    >>> print(len(module_doc) > 0)
    True

    >>> funciont_doc = __import__("101-lazy_matrix_mul").lazy_matrix_mul.__doc__
    >>> print(len(funciont_doc) > 0)
    True

General usege
==============

    >>> a = [[3, 2, 1, 5], [9, 1, 3, 0]]
    >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
    >>> lazy_matrix_mul(a, b)
    array([[50, 42, 42],
           [25, 96, 26]])

	>>> a = [[1, 2],[2, 3], [1, 2],[2, 3],[1, 2],[2, 3]]
	>>> b = [[1, 2, 3],[1, 2, 3]]
	>>> lazy_matrix_mul(a, b)
	array([[ 3,  6,  9],
	       [ 5, 10, 15],
	       [ 3,  6,  9],
	       [ 5, 10, 15],
	       [ 3,  6,  9],
	       [ 5, 10, 15]])

Handling Erros
==============
Handling not allowed arguments

	 >>> a = 2
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: Scalar operands are not allowed, use '*' instead

	 >>> b = "string"
	 >>> a = [[2, 39, 0], [1, 3, 32], [2, 24, 7], [8, 18, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: Scalar operands are not allowed, use '*' instead

	 >>> a = []
	 >>> b = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (0,) and (2,2) not aligned: 0 (dim 0) != 2 (dim 0)

	 >>> b = []
	 >>> a = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

	 >>> b = []
	 >>> a = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

	 >>> b = [[]]
	 >>> a = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (2,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

	 >>> a = [[]]
	 >>> b = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (1,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

	 >>> a = [[1, 2.456],[2, 3.27386]]
	 >>> b = [[1, 2],[3, "6"]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: invalid data type for einsum

	 >>> a = [[1, 2.456],[2, 3], [32]]
	 >>> b = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> b = [[1, 2.456],[2, 3], [24]]
	 >>> a = [[1, 2],[3, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> a = [[1, 2],[3, 4]]
	 >>> b = [[1, 2, 3],[4, 2, 3],[1, 6, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (2,2) and (3,3) not aligned: 2 (dim 1) != 3 (dim 0)

	 >>> b = True
	 >>> a = [[1, 2, 3],[1, 2, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: Scalar operands are not allowed, use '*' instead

	 >>> a = True
	 >>> b = [[1, 2, 3],[1, 2, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: Scalar operands are not allowed, use '*' instead

	 >>> b = None
	 >>> a = [[1, 2, 3],[1, 2, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: Object arrays are not currently supported

	 >>> b = 23
	 >>> a = [[1, 2, 3],[1, 2, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: Scalar operands are not allowed, use '*' instead

	 >>> a = [[1, 2],[3, 4]]
	 >>> b = [[1, 2, 3],[4, 2, 3],[1, 6, 3]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: shapes (2,2) and (3,3) not aligned: 2 (dim 1) != 3 (dim 0)

	 >>> a = [[1, 2], 32]
	 >>> b = [[2, 9, 32], [1, 6, 5], [4, 4, 7], [8, 3, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> b = [[1, 2], 32]
	 >>> a = [[2, 9, 32], [1, 6, 5], [4, 4, 7], [8, 3, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, "3"]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: invalid data type for einsum

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, None]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: Object arrays are not currently supported

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, [1, 3, 4]]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, {"hola"}]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: Object arrays are not currently supported

	 >>> a = [[3, 2, "33", 5], [9, 1, 3, "3"]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 1, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: invalid data type for einsum

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, "hello", 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: invalid data type for einsum

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, [1, 3, 4], 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, {"hello"}, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 TypeError: Object arrays are not currently supported

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, False, 5]]
	 >>> lazy_matrix_mul(a, b)
	 array([[50, 37, 42],
	        [49, 96, 41]])

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 3, 5, 4]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> a = [[3, 2, 1, 5], [9, 1, 3, 3, 4]]
	 >>> b = [[2, 9, 0], [1, 3, 5], [2, 4, 7], [8, 3, 5]]
	 >>> lazy_matrix_mul(a, b)
	 Traceback (most recent call last):
	 ValueError: setting an array element with a sequence.

	 >>> lazy_matrix_mul([[1, 1]], [[2, 2], [2, 3]], [[3, 2], [3, 2]])
	 Traceback (most recent call last):
	 TypeError: lazy_matrix_mul() takes 2 positional arguments but 3 were given

	 >>> lazy_matrix_mul([[1, 1]])
	 Traceback (most recent call last):
	 TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

	 >>> lazy_matrix_mul([ ,[1, 1]])
	 Traceback (most recent call last):
	 SyntaxError: invalid syntax

	 >>> lazy_matrix_mul()
	 Traceback (most recent call last):
	 TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

	 >>> lazy_matrix_mul([[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]])
	 Traceback (most recent call last):
	 TypeError: lazy_matrix_mul() takes 2 positional arguments but 3 were given

	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1.5, 2.5, 3.5], [4.5, 6.5, 7.5]])
	 array([[10.5, 15.5, 18.5],
		[22.5, 33.5, 40.5]])

	 >>> lazy_matrix_mul([[], []], [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ValueError: shapes (2,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

Test 2 bad matrices:

    >>> lazy_matrix_mul([[1, 2, 3], [3, 4, 5]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
    ...
    ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

Test 2 ok matrices:

    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    array([[19, 22,  5],
	   [43, 50, 11],
	   [43, 50, 11]])

Test singly empty matrix:

    >>> lazy_matrix_mul([], [[3]])
    Traceback (most recent call last):
    ValueError: shapes (0,) and (1,1) not aligned: 0 (dim 0) != 1 (dim 0)

Test singly empty matrix:

    >>> lazy_matrix_mul([[]], [[3]])
    Traceback (most recent call last):
    ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)

Test singly empty matrix 2:

    >>> lazy_matrix_mul([[4]], [])
    Traceback (most recent call last):
    ValueError: shapes (1,1) and (0,) not aligned: 1 (dim 1) != 0 (dim 0)

Test doubly empty matrix:

    >>> lazy_matrix_mul([[3]], [[]])
    array([], shape=(1, 0), dtype=int64)

Test simple case:

    >>> lazy_matrix_mul([[3]], [[9]])
    array([[27]])

Test string arg:

    >>> lazy_matrix_mul("foo", [[9]])
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

Test string arg:

    >>> lazy_matrix_mul([[8]], "bar")
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead


Test int list:

    >>> lazy_matrix_mul([1, 3, 4], [[9]])
    Traceback (most recent call last):
    ValueError: shapes (3,) and (1,1) not aligned: 3 (dim 0) != 1 (dim 0)

Test int list 2:

    >>> lazy_matrix_mul([[8]], [2, 4, 6])
    Traceback (most recent call last):
    ValueError: shapes (1,1) and (3,) not aligned: 1 (dim 1) != 3 (dim 0)

Test notnum list:

    >>> lazy_matrix_mul([["foo"]], [[2, 4, 6]])
    Traceback (most recent call last):
    TypeError: invalid data type for einsum

Test notnum list 2 :

    >>> lazy_matrix_mul([[2, 4, 6]], [["foo"]])
    Traceback (most recent call last):
    ValueError: shapes (1,3) and (1,1) not aligned: 3 (dim 1) != 1 (dim 0)

Test a bad rows:

    >>> lazy_matrix_mul([[1, 2], [3, 4, 5], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

Test b bad rows:

    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 2]])
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

Test missing 1 arg:

    >>> lazy_matrix_mul([[3]])
    Traceback (most recent call last):
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

Test missing 2 args:

    >>> lazy_matrix_mul()
    Traceback (most recent call last):
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
